### PR TITLE
Adding step_geohash

### PR DESCRIPTION
**Pull request by jrosell**
Related to https://github.com/JosiahParry/geohash/issues/1

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
 export(decode)
 export(decode_bbox)
 export(encode)
 export(encode_par)
 export(neighbor)
 export(neighbors)
+export(step_geohash)
+import(recipes)
+import(rlang)
+importFrom(cli,cli_abort)
```
Comment (JosiahParry) [inline NAMESPACE]:
```suggestion
```

File: DESCRIPTION [hunk 2]
```diff
@@ -1,14 +1,22 @@
...
 Language: en
 Roxygen: list(markdown = TRUE)
-RoxygenNote: 7.3.1
+RoxygenNote: 7.3.2
 Config/rextendr/version: 0.3.1.9001
 SystemRequirements: Cargo (Rust's package manager), rustc >= 1.70
+Imports: 
+    cli,
+    recipes,
+    rlang (>= 1.1.0)
```
Comment (JosiahParry) [inline DESCRIPTION]: It is really important to me that we keep the package dependency free. These should all be moved to suggests

File: DESCRIPTION [hunk 3]
```diff
@@ -1,14 +1,22 @@
...
+Suggests:
+    testthat (>= 3.0.0)
```
Comment (JosiahParry) [inline DESCRIPTION]:
```suggestion
Suggests:
    cli,
    recipes,
    rlang (>= 1.1.0),
    testthat (>= 3.0.0)
```

File: R/geohash-package.R [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#' @keywords internal
+"_PACKAGE"
+
+## usethis namespace: start
+#' @importFrom cli cli_abort
+#' @import recipes
+#' @import rlang
```
Comment (JosiahParry) [inline R/geohash-package.R]:
```suggestion
```

File: R/step_geohash.R [hunk 5]
```diff
@@ -0,0 +1,143 @@
+
```
Comment (JosiahParry) [inline R/step_geohash.R]: While annoying, I think we can recreate the `check_string()` and `check_type()` by fetching them from the namespace instead 

```r
check_string <- getFromNamespace("check_string", "recipes")
check_type <- getFromNamespace("check_type", "recipes")
```

File: R/import-standalone-types-check.R
Comment (JosiahParry) [inline R/import-standalone-types-check.R:1]: I believe we can delete this entire file

File: R/import-standalone-obj-type.R
Comment (JosiahParry) [inline R/import-standalone-obj-type.R:1]: i believe we can delete this entire file

File: R/step_geohash.R [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  lat = NULL,
+  role = NA,
+  trained = FALSE,
+  options = list(length = 8, parallel = FALSE),
+  name = "geohash",
+  columns = NULL,
+  skip = FALSE,
+  id = rand_id("geohash")
+) {
+  check_string(name)
```
Comment (JosiahParry) [inline R/step_geohash.R]:
```suggestion
) {

   if (!requireNamespace("recipes")) {
       stop("The package `recipes` is required for this functionality")
   }
   
  check_string <- getFromNamespace("check_string", "recipes")

  check_string(name)
```

File: R/step_geohash.R [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  role,
+  trained,
+  options,
+  name,
+  columns,
+  skip,
+  id
+) {
+  step(
+    subclass = "geohash",
```
Comment (JosiahParry) [inline R/step_geohash.R]:
```suggestion
) {
     if (!requireNamespace("recipes")) {
       stop("The package `recipes` is required for this functionality")
   }
  step(
    subclass = "geohash", 
```

File: R/step_geohash.R (hunk 5 above)
Comment (jrosell) [reply R/step_geohash.R]: Oh. Great, this is why I added deps.

File: R/step_geohash.R [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+#' @export
+prep.step_geohash <- function(
+  x,
+  training,
+  info = NULL,
+  ...
+) {
+  lon_name <- recipes_eval_select(x$lon, training, info)
+  lat_name <- recipes_eval_select(x$lat, training, info)
```
Comment (JosiahParry) [inline R/step_geohash.R]:
```suggestion
   if (!requireNamespace("recipes")) {
       stop("The package `recipes` is required for this functionality")
   }
  check_type <- getFromNamespace("check_type", "recipes")
  lon_name <- recipes::recipes_eval_select(x$lon, training, info)
  lat_name <- recipes::recipes_eval_select(x$lat, training, info)
```

File: R/step_geohash.R [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    columns = c(lon_name, lat_name),
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+#' @export
+bake.step_geohash <- function(object, new_data, ...) {
+  col_names <- names(object$columns)
+  check_new_data(col_names, object, new_data)
```
Comment (JosiahParry) [inline R/step_geohash.R]:
```suggestion
bake.step_geohash <- function(object, new_data, ...) {
   if (!requireNamespace("recipes")) {
       stop("The package `recipes` is required for this functionality")
   }
   
  col_names <- names(object$columns)
  recipes::check_new_data(col_names, object, new_data)
```

File: tests/testthat/test-step_geohash.R [hunk 10]
```diff
@@ -0,0 +1,23 @@
+test_that("step_geohash works", {
+  n <- 10
+  df <- data.frame(
+    y = rnorm(n, 0, 1),
+    longitude  = runif(n, -180, 180),
+    latitude = runif(n, -90, 90)
```
Comment (JosiahParry) [inline tests/testthat/test-step_geohash.R:8]:
```suggestion
  testthat::skip_if_not_installed("recipes")
  n <- 10
  df <- data.frame(
    y = rnorm(n, 0, 1),
    longitude  = runif(n, -180, 180),
    latitude = runif(n, -90, 90)
```

**Comment by JosiahParry** [comment]
Thank you for the PR! This looks lovely! I bloodied it up a bit! But I _think_ I made all of the suggestions that are needed! 

The important thing to me is to keep this dependency free. So the recipes bit needs to be gated behind a `requireNamespace()` check. 

Fortunately, recipes contains all of the checks that we need so we can yank them from their namespace and use them. 

I've also added `recipes::` in a few places where I think is necessary as well as a `skip_if_not_installed()` in the testthat.

Would you mind going through the suggestions and see what I may have missed? 

Super exciting!!

**Comment by jrosell** [comment]
Check it now. :)

**Comment by JosiahParry** [comment]
Thanks! I think a follow up PR should rename the arguments from `lon` and `lat` to their full names to match other function signatures.

**Comment by jrosell** [comment]
Idk, keep in mind I followed step_geodist argument names as convention :)

**Comment by JosiahParry** [comment]
All good! {geodist} has `x` and `y` arguments. A `geohash` is only available for coordinates that are specified in latitude and longitude units  WGS 84 / EPSG:4326 which must be in the range of [-180, 180] and [-90, 90]. geodist permits you to calculate distance metrics for any spherical object. 

Being as explicit as possible is always better imo

**Comment by jrosell** [comment]
Good take. I wonder if some input checking & warning/abort should be done or return NA.

**Pull request merged**